### PR TITLE
distsqlrun: fix double output close in joinReader

### LINKED ISSUE #14360: distsql: panic of TPC-H query 3 when running on 6-node omega
Running TPC-H Query 3:

```
SELECT
    l_orderkey,
    SUM(l_extendedprice * (1 - l_discount)) AS revenue,
    o_orderdate,
    o_shippriority
FROM
    customer,
    orders,
    lineitem
WHERE
    c_mktsegment = 'MACHINERY'
    AND c_custkey = o_custkey
    AND l_orderkey = o_orderkey
    AND o_orderDATE < DATE '1995-03-10'
    AND l_shipdate > DATE '1995-03-10'
GROUP BY
    l_orderkey,
    o_orderdate,
    o_shippriority
ORDER BY
    revenue DESC,
    o_orderdate;
```

on 6-node `omega` with sha `78c5f93` results in:
```
E170324 14:13:25.364501 3713 sql/distsqlrun/server.go:216  [n1] [n1] communication error: rpc error: code = Canceled desc = context canceled
E170324 14:13:25.364788 3992 sql/distsqlrun/server.go:216  [n1] [n1] communication error: rpc error: code = Canceled desc = context canceled
panic: close of closed channel

goroutine 4098 [running]:
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*RowChannel).ProducerDone(0xc42390a630)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:320 +0x2f
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*procOutputHelper).close(0xc4226d25f8)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:280 +0x34
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*joinReader).Run(0xc4226d0a00, 0x7f326ac11040, 0xc421241e30, 0xc422a464a8)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/joinreader.go:199 +0x1bf
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:309 +0x337
```

This is caused by the same undesired behavior seen in #13989 but this issue tracks the panic caused by it. The execution plan can be found [here](https://raduberinde.github.io/decode.html?eJzclN9v2jAQx9_3V1j30rK5Ek4CdJYqGbVszTRgAvYwTVGV4htECjGyHWlVxf8-YbYSUmTW0UrV3i62P3ff-5G7h0JJHKQLNMC_AwMKESQUllpN0Ril18ebR7H8CbxJISuWpV0fJxSmSiPwe7CZzRE4fFJZMcJUogYKEm2a5c7vUmeLVN8JpSVqAxSGpeVEMCoCKlpUnEOyoqBK-8f1ij5f0DwrMLO4qIZtU9H565jbUGXhMkC5E8mRh57sEX6dmvlafF34cEBy_GFPBWtc6Gw2d9aD-oCKkIrI1W1PGlv3t3dknpr5rm_BIFm9_lRHWEjUnIg2ER3ylpwyzvlV7zLudz-TMyLOG0RERLTcg7VxzAB1ZzONs9SqmriHClOIr3qDyW9t46_9UxE0tqdhxY4qdqtityt2p_GMo_fy-RxR2bHStt5yEZxREb7zumV-t5P0NkffLz8tjVULd_shy62bpA65ICf97uV1POiNvp1wzseTUTz4WF0Lx_9M_6pc3Sj5aD02X6RI-Y2Re5aifx-y_35JHCjnK14Se9SO0CxVYbDelx1_CQWUM9y01KhST_GLVlPnfPM5dK_dgURjN7fB5iMu3BVby6rC7AlwUIeDOsyqMGseQ4d-OPTqjnbgZh2OvHDLX7GWF277I7e9MAv8dOcp9X4k3E-HfvjcK7zp1_3en7UfZv75Zodw_4TXm52s3vwKAAD__2AOoVo=).

The same thing does not happen on a local 3-node cluster [(execution plan)](https://raduberinde.github.io/decode.html?eJzUlF9r2zwUxu_fTyHOTZN3KsR2_k1QUGiz1WNJRpJdjGGKZ525BscKkgwbJd99WN6axGFaE8Ngd5KOfud59FjyExRS4DzeoAb2GTyg4AOFACIKWyUT1FqqqlRvDMU3YD0KWbEtTbUcUUikQmBPYDKTIzBYx19yXGIsUAEFgSbOctt8q7JNrL7zpNRGbmz1TZYbVIzwEbkhV7PJ7X04ny4_XTHGVutlOH8LFBalYYR7lI8g2lGQpfmlvaMvd_VOZoXLlFQClT6U8ykfUD5uI_rbKPIHLXieFZgZ3OxFX37CvUZZWOsojhSi3QUZnBqifHhO7n92RS8yfh_rx8p80_hiTnL8ajrc696oLH20o2f3PuUB5X37Gf_JYyyxEPZ1DKsH8j_peIyxu-ltOJu8J9eEj7uE9wkf2A3VoM1dnaSpwjQ2smHuOT0K4d10vv7pbfVx1uF-d78aHIz73TZOVlKZZkTcv6Y8eOVs6_3Nx-hfKCYfpDj52fTaxLVEvZWFxuYVPeoXUUCRYn27tSxVgh-UTGzzerqwu-2CQG3q6qCehIUteZWtQ9hzwn037DvhwA0HZyj7Tbh_xplP4IETHh7BvSY8dMIjNzxywl7PTY_bxP3aLd1Qjnb__QgAAP__K9O2Jw==). I assume it is because the `JoinReader`s read from two different nodes rather than one, as is the case with `omega`.

### PR DESCRIPTION
**Pull request by asubiotto**
Fixes #14360.

We had a bug when emitHelper() returned false - because the consumer was
closed. A ret val of false means that emitHelper() also closed the
output. But the joinReader was trying to close it again, resulting in a
panic.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/14420)

### CODE REVIEW

File: pkg/sql/distsqlrun/joinreader.go [hunk 1]
```diff
@@ -134,6 +135,7 @@ func (jr *joinReader) mainLoop(ctx context.Context) error {
 			}
 			if row == nil {
 				if len(spans) == 0 {
+					jr.out.close()
```
Comment (vivekmenezes) [inline pkg/sql/distsqlrun/joinreader.go]: Are we able to use a defer here based on the value of jr.out != nil ?

**Comment by asubiotto** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/distsqlrun/joinreader.go, line 138 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420#-KgKaghy11RAvGksVMik-r1-138:-KgQ4UvALQzTzNgF3FmJ:b-nqev6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/e690775ef3140850e4120c2bbea6b61deff44065/pkg/sql/distsqlrun/joinreader.go#L138)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Are we able to use a defer here based on the value of jr.out != nil ?
</blockquote></details>

No, jr.out is a value. I guess we could do a similar thing but other code follows this design and it seems a bit weird to me to add a check like that. Maybe we have to rethink the design of emitHelper and where the closing responsibility lies.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420)*

**Comment by cuongdo** [comment]
@andreimatei any thoughts here? our TPC-H load generator is hitting this very quickly, which makes quality load testing hard

**Comment by andreimatei** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/distsqlrun/joinreader_test.go, line 174 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420#-KgQUMekCwmX1OlEAzgb:-KgQUMekCwmX1OlEAzgc:bplzpuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e690775ef3140850e4120c2bbea6b61deff44065/pkg/sql/distsqlrun/joinreader_test.go#L174)):*
> ```Go
> 
> 	in := &RowBuffer{}
> 	encRow := make(sqlbase.EncDatumRow, numRows)
> ```

`numRows` here seems wrong. You want one *column*.

---

*[pkg/sql/distsqlrun/joinreader_test.go, line 184 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420#-KgQVUrdU47ZBrHcRig3:-KgQVUrdU47ZBrHcRig4:b4sitsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/e690775ef3140850e4120c2bbea6b61deff44065/pkg/sql/distsqlrun/joinreader_test.go#L184)):*
> ```Go
> 
> 	out := &RowBuffer{}
> 	out.ConsumerClosed()
> ```

I'm confused about how the test passes.
`out.ConsumerClosed()` should cause `emitHelper` (when called by the jr) to call `ConsumerClosed()` on `in`. And that call doesn't set `in.Done`.
So technically with the way the test is written, we shouldn't be expecting any "draining" (so the fatal message below is wrong). We should simply expect that `in` is closed (it has some status field).
If you want to test draining, you should call `out.ConsumerDone()`.
Does this make sense?

And yet the test presumably passes, so I'm missing something.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420)*

**Comment by andreimatei** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/distsqlrun/joinreader_test.go, line 184 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420#-KgQVUrdU47ZBrHcRig3:-KgQYUtP_R93pIDxcjtR:b-hhxdih) ([raw file](https://github.com/cockroachdb/cockroach/blob/e690775ef3140850e4120c2bbea6b61deff44065/pkg/sql/distsqlrun/joinreader_test.go#L184)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'm confused about how the test passes.
`out.ConsumerClosed()` should cause `emitHelper` (when called by the jr) to call `ConsumerClosed()` on `in`. And that call doesn't set `in.Done`.
So technically with the way the test is written, we shouldn't be expecting any "draining" (so the fatal message below is wrong). We should simply expect that `in` is closed (it has some status field).
If you want to test draining, you should call `out.ConsumerDone()`.
Does this make sense?

And yet the test presumably passes, so I'm missing something.

</blockquote></details>

oh, the `jr` tries to read a batch of rows from `in`. So it will set `in.Done` even before it tries to emit anything.
So we're fooling ourselves into thinking that the joiner drained anything.

You can put more than the batch size rows in the buffer to avoid this.
Better yet, I'd try to add the ability to our row buffer to generate an infinite sequence of rows. The `RowBuffer` is already a mock object for tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420)*

**Comment by asubiotto** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/distsqlrun/joinreader_test.go, line 174 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420#-KgQUMekCwmX1OlEAzgb:-KgUWKVNn1CBQtzMyagK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e690775ef3140850e4120c2bbea6b61deff44065/pkg/sql/distsqlrun/joinreader_test.go#L174)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`numRows` here seems wrong. You want one *column*.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/joinreader_test.go, line 184 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420#-KgQVUrdU47ZBrHcRig3:-KgUTHqe6bAq0Us1BUbn:bbsuwbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e690775ef3140850e4120c2bbea6b61deff44065/pkg/sql/distsqlrun/joinreader_test.go#L184)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

oh, the `jr` tries to read a batch of rows from `in`. So it will set `in.Done` even before it tries to emit anything.
So we're fooling ourselves into thinking that the joiner drained anything.

You can put more than the batch size rows in the buffer to avoid this.
Better yet, I'd try to add the ability to our row buffer to generate an infinite sequence of rows. The `RowBuffer` is already a mock object for tests.
</blockquote></details>

Got it, thanks for explaining. Switched to check the `ConsumerStatus` of `in`. Since this test doesn't really care about the producer apart from the fact that it closed, I'll leave it at one row.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420)*

**Comment by vivekmenezes** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/joinreader.go, line 138 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420#-KgKaghy11RAvGksVMik-r1-138:-KgUYcqroJBGc4YasNp-:b9qbzdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e690775ef3140850e4120c2bbea6b61deff44065/pkg/sql/distsqlrun/joinreader.go#L138)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

No, jr.out is a value. I guess we could do a similar thing but other code follows this design and it seems a bit weird to me to add a check like that. Maybe we have to rethink the design of emitHelper and where the closing responsibility lies.
</blockquote></details>

Is it possible for the two points to break out and then for the close() to happen depending on a condition on the len(spans). You can document in one place why you're calling close()

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420)*

**Comment by asubiotto** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/distsqlrun/joinreader.go, line 138 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420#-KgKaghy11RAvGksVMik-r1-138:-KgUjA3FXkJbtUPPDM49:b-99585v) ([raw file](https://github.com/cockroachdb/cockroach/blob/e690775ef3140850e4120c2bbea6b61deff44065/pkg/sql/distsqlrun/joinreader.go#L138)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Is it possible for the two points to break out and then for the close() to happen depending on a condition on the len(spans). You can document in one place why you're calling close()
</blockquote></details>

Added a comment as to why this is being closed here. As discussed offline, this solution is fine for now but I will think of a way to make the general closing methodology cleaner (since we have a repeat of this kind of code in the other processors too).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420)*

**Comment by andreimatei** [comment]
:lgtm:

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/joinreader_test.go, line 184 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420#-KgQVUrdU47ZBrHcRig3:-KgUm2bCfL9hzDKSi0oS:bvmmb5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/e690775ef3140850e4120c2bbea6b61deff44065/pkg/sql/distsqlrun/joinreader_test.go#L184)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Got it, thanks for explaining. Switched to check the `ConsumerStatus` of `in`. Since this test doesn't really care about the producer apart from the fact that it closed, I'll leave it at one row.
</blockquote></details>

well technically if the producer has been exhausted, calling `ConsumerDone` and/or `ConsumerClosed` on it is not required. So the `joinReader` is free to not call `ConsumerClosed()` on `in`. 

Consider also adding a (sub-test) for draining (so call `out.ConsumerDone`) above and see that some metadata coming from `in` is still read.
We should also leave the test with the `out.ConsumerClosed` call and see that we don't crash (and whether or not `in.ConsumerStatus` is tested or not, I don't care).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420#-:-KgUnZSQWKCIgXFqEjzB:bnfp4nl)*

**Comment by asubiotto** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/joinreader_test.go, line 184 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420#-KgQVUrdU47ZBrHcRig3:-KgVhB1osxOH0A0arQg_:bs18207) ([raw file](https://github.com/cockroachdb/cockroach/blob/e690775ef3140850e4120c2bbea6b61deff44065/pkg/sql/distsqlrun/joinreader_test.go#L184)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

well technically if the producer has been exhausted, calling `ConsumerDone` and/or `ConsumerClosed` on it is not required. So the `joinReader` is free to not call `ConsumerClosed()` on `in`. 

Consider also adding a (sub-test) for draining (so call `out.ConsumerDone`) above and see that some metadata coming from `in` is still read.
We should also leave the test with the `out.ConsumerClosed` call and see that we don't crash (and whether or not `in.ConsumerStatus` is tested or not, I don't care).
</blockquote></details>

Done. Added the sub-test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14420)*

**Pull request merged**